File: js/game/card.js
// SMØR - Card-klasse (konvertert fra game/card.py)

class Card {
  constructor(
    name,
    promilleChange = 0,
    minneChange = 0,
    specialEffect = null,
    displayText = "",
    idName = null,
    onlyName = null,
    icon = null,
    shortEffect = null
  ) {
    this.name = name;
    this.idName = idName || name; // Use idName if provided, otherwise fall back to name
    this.onlyName = onlyName || name; // Use onlyName if provided, otherwise fall back to name
    this.icon = icon || ""; // Use icon if provided, otherwise empty string
    this.shortEffect = shortEffect || ""; // Use shortEffect if provided, otherwise empty string
    this.promilleChange = promilleChange;
    this.minneChange = minneChange;
    this.specialEffect = specialEffect; // spesialeffekt som "call_friend"
    this.displayText = displayText;
  }

  async play(player, place, gameEngine = null) {
    // Håndter chugging først for øl-kort (før øl-effekten gis)
    if (this.idName === "drinkBeer" && gameEngine) {
      await gameEngine.handleBeerChug(player);
    }

    // Håndter chugging for ice-kort (før ice-effekten gis)
    if (this.idName === "drinkIce" && gameEngine) {
      await gameEngine.handleIceChug(player);
    }

    // Endre promille og minner (øl-effekten gis alltid)
    player.addPromille(this.promilleChange);
    player.addMemory(this.minneChange);
    player.lastDrinkCard = this.idName;
    player.lastCardPromille = this.promilleChange;

    // Sjekk om stedet gir ekstra effekt FØRST
    const currentPhase = gameEngine ? gameEngine.currentPhase : null;
    await place.applyPhaseEffect(player, currentPhase, gameEngine);

    if (gameEngine) {
      // Check if this beer card becomes a drink card due to place effect
      let cardNameForLog = this.name;
      if (this.idName === "drinkBeer" && place.effects.beer_as_drink) {
        cardNameForLog = "Drikk en drink";
      }

      let effectText = "";
      if (this.promilleChange !== 0) {
        effectText += `Promille: ${this.promilleChange > 0 ? "+" : ""}${
          this.promilleChange
        }%`;
      }

      // Check if this is a beer card that became a drink on Oslo Plaza
      let actualMinneChange = this.minneChange;
      if (this.idName === "drinkBeer" && place.effects.beer_as_drink) {
        actualMinneChange = 1; // Drink gives 1 memory
      }

      if (actualMinneChange !== 0) {
        if (effectText) effectText += ", ";
        effectText += `M: ${
          actualMinneChange > 0 ? "+" : ""
        }${actualMinneChange}`;
      }

      if (effectText) {
        await gameEngine.addToLog(
          `${player.name} spiller ${cardNameForLog} -> ${effectText}`,
          "info"
        );
      }
    }

    // Håndter spesialeffekter
    if (this.specialEffect === "call_friend" && gameEngine) {
      await gameEngine.addToLog(`📞 RING EN VENN`, "info");
      await gameEngine.offerNpcInteraction(player);
    } else if (this.specialEffect === "know_beer" && gameEngine) {
      // For know_beer, check NPC effects first to get enhanced effect
      let enhancedEffect = null;
      for (const npc of player.npcs) {
        const result = await npc.applyCardEffects(player, this.name);
        if (result) {
          enhancedEffect = result;
        }
      }
      await gameEngine.handleKnowBeerEffect(player, enhancedEffect);
    } else if (this.specialEffect === "bong_choice" && gameEngine) {
      await gameEngine.handleBongChoice(player);
    } else if (this.specialEffect === "round_drinks" && gameEngine) {
      await gameEngine.handleRoundDrinks(player);
    } else if (this.specialEffect === "icing" && gameEngine) {
      // Set flag to prevent main.js from setting waitingForHumanInput = false
      gameEngine.waitingForIcing = true;
      await gameEngine.handleIcing(player);
      // Set both flags to false after icing is complete
      gameEngine.waitingForIcing = false;
      gameEngine.waitingForHumanInput = false;
      return; // Exit early, special effect will handle the flow
    } else if (this.specialEffect === "call_friend" && gameEngine) {
      await gameEngine.offerNpcInteraction(player);
      // Don't set waitingForHumanInput = false here - let the special effect handle it
      return; // Exit early, special effect will handle the flow
    } else {
      // For other cards, check NPC effects after special effects
      let enhancedEffect = null;
      for (const npc of player.npcs) {
        const result = await npc.applyCardEffects(player, this.name);
        if (result) {
          enhancedEffect = result;
        }
      }
    }

    // Ensure game continues after card is fully played
    if (gameEngine) {
      gameEngine.waitingForHumanInput = false;
    }
  }
}

File: game/card.py
# -*- coding: utf-8 -*-
# card.py

class Card:
    def __init__(self, name, promille_change=0, minne_change=0, special_effect=None, display_text=""):
        self.name = name
        self.promille_change = promille_change
        self.minne_change = minne_change
        self.special_effect = special_effect  # spesialeffekt som "call_friend"
        self.display_text = display_text

    def play(self, player, place, game_engine=None):
        # Håndter chugging først for øl-kort
        if self.name == "Drikk en øl" and game_engine:
            game_engine.handle_beer_chug(player)
        
        # Endre promille og minner
        player.add_promille(self.promille_change)
        player.add_memory(self.minne_change)
        player.last_drink_card = self.name
        player.last_card_promille = self.promille_change

        print(f"{player.name} spiller {self.name} -> Promille: {player.promille}, Minne: {player.memory}")

        # Sjekk om stedet gir ekstra effekt
        current_phase = game_engine.current_phase if game_engine else None
        place.apply_phase_effect(player, current_phase)
        
        # Sjekk NPC-effekter
        enhanced_effect = None
        for npc in player.npcs:
            result = npc.apply_card_effects(player, self.name)
            if result:
                enhanced_effect = result
        
        # Håndter spesialeffekter
        if self.special_effect == "call_friend" and game_engine:
            game_engine.offer_npc_interaction(player)
        elif self.special_effect == "know_beer" and game_engine:
            game_engine.handle_know_beer_effect(player, enhanced_effect)
        elif self.special_effect == "bong_choice" and game_engine:
            game_engine.handle_bong_choice(player)
        elif self.special_effect == "round_drinks" and game_engine:
            game_engine.handle_round_drinks(player)
        elif self.special_effect == "icing" and game_engine:
            game_engine.handle_icing(player)
